common/store.go

```diff
@@ -0,0 +1,265 @@
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+package common
+
+import (
+	"errors"
+	"fmt"
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/jinzhu/gorm"
+	_ "github.com/mattn/go-sqlite3"
+	"log"
+	"os"
+	"strconv"
+)
+
+// MultiError adapts GORM (ORM - see https://github.com/jinzhu/gorm) array of errors found in GetErrors()
+// to a single error interface.
+// GORMdoes not return errors at every turn. It accumulates them and returns
+// them whenever you feel like calling GetErrors() (https://godoc.org/github.com/jinzhu/gorm#DB.GetErrors).
+// Since this is not consistent  with the rest of the code, I prefer to isolate it
+// here and make an adapter.
+type MultiError struct {
+	errors []error
+}
+
+func MakeMultiError(errors []error) error {
+	if errors == nil {
+
+		return nil
+	}
+	if len(errors) == 0 {
+
+		return nil
+	}
+
+	return &MultiError{errors}
+}
+
+func (m *MultiError) Error() string {
+	s := ""
+	for i := range m.errors {
+		if len(s) > 0 {
+			s += "; "
+		}
+		s += m.errors[i].Error()
+	}
+	return s
+}
+
+// Stores information needed for a DB connection.
+type StoreConfig struct {
+	Host     string
+	Port     uint64
+	Username string
+	Password string
+	Database string
+	Type     string
+}
+
+// MakeStoreConfig creates StoreConfig object from a map.
+func makeStoreConfig(configMap map[string]interface{}) StoreConfig {
+	storeConfig := StoreConfig{}
+	storeConfig.Type = configMap["type"].(string)
+	if configMap["host"] != nil {
+		storeConfig.Host = configMap["host"].(string)
+	}
+	if configMap["port"] != nil {
+		portStr := configMap["port"].(string)
+		port, err := strconv.ParseUint(portStr, 10, 64)
+		if err != nil {
+			log.Printf("Error parsing %s", portStr)
+		} else {
+			storeConfig.Port = port
+		}
+	}
+	if configMap["username"] != nil {
+		storeConfig.Username = configMap["username"].(string)
+	}
+	if configMap["password"] != nil {
+		storeConfig.Password = configMap["password"].(string)
+	}
+	storeConfig.Database = configMap["database"].(string)
+	return storeConfig
+}
+
+// Defines generic store interface that can be used
+// by any service for persistence.
+type Store interface {
+	// SetConfig sets the configuration
+	SetConfig(configMap map[string]interface{}) error
+	// Connect connects to the store
+	Connect() error
+	// Create the schema, dropping existing one if the true flag is specified
+	CreateSchema(force bool) error
+}
+
+// ServiceStore interface is what each service's store needs to implement.
+type ServiceStore interface {
+	// Entities returns list of entities (DB tables) this store is managing.
+	Entities() []interface{}
+	// CreateSchemaPostProcess runs whatever required post-processing after
+	// schema creation (perhaps initializing DB with some initial or sample data).
+	CreateSchemaPostProcess() error
+}
+
+// Function implementing createSchema functionality
```

Comment:
"createSchema is a type for functions that create database schemas. By defining a type we can more easily store references to functions of the specified signature."